### PR TITLE
Remove OfferedDeadlineWatchdog

### PR DESCRIPTION
**Pull request by jrw972**
The watchdog uses timer_ids and therefore must wait() when scheduling
timers.  This effectively introduces a lock which leads to deadlock
when the Deadline QoS is used.  The deadly embrace occurs between the
writing thread which has the DataWriterImpl lock and needs the reactor
lock and the Reactor thread servicing the watchdog which has the
reactor thread and needs the DataWriterImpl lock.

Solution: Rewrite the OfferedDeadlineWatchdog using SporadicTask.

In general, wait()ing on Command objects in the ReactorInterceptor is
unsafe and the wait functionality should be removed to reduce the
chance of similar deadlocks in the future.

### CODE REVIEW

File: dds/DCPS/WriteDataContainer.h [hunk 1]
```diff
@@ -116,7 +119,7 @@ typedef OPENDDS_MAP(DDS::InstanceHandle_t, PublicationInstance_rch)
  *           we do not deadlock; and, 2) we incur the cost of
  *           obtaining the lock only once.
  */
-class OpenDDS_Dcps_Export WriteDataContainer : public PoolAllocationBase {
+class OpenDDS_Dcps_Export WriteDataContainer : public RcObject {
```
Comment (jrw972) [inline dds/DCPS/WriteDataContainer.h:122]: @mitza-oci We need to discuss the ramifications of changing this to RcObject.

File: dds/DCPS/WriteDataContainer.cpp [hunk 2]
```diff
@@ -1500,6 +1488,159 @@ WriteDataContainer::log_send_state_lists (OPENDDS_STRING description)
...
+        deadline_task_->schedule(deadline_period);
+      }
+    } else if (deadline_period == TimeDuration::max_value) {
+      if (!deadline_map_.empty()) {
+        deadline_task_->cancel();
+      }
+
+      deadline_map_.clear();
+    } else {
+      // TODO: Which is correct?  The first can cause immediate deadline misses?  The second corresponds to the existing behavior.
```
Comment (jrw972) [inline dds/DCPS/WriteDataContainer.cpp:1520]: @mitza-oci Another discussion point.

File: dds/DCPS/WriteDataContainer.h [hunk 3]
```diff
@@ -528,6 +531,28 @@ class OpenDDS_Dcps_Export WriteDataContainer : public PoolAllocationBase {
   DDS::DurabilityServiceQosPolicy const & durability_service_;
 
 #endif
+
+  /// Timer responsible for reporting missed offered deadlines.
+  RcHandle<DCPS::PmfSporadicTask<WriteDataContainer> > deadline_task_;
+  TimeDuration deadline_period_; // TimeDuration::zero_value means no deadline.
+  typedef std::multimap<MonotonicTimePoint, PublicationInstance_rch> DeadlineMapType;
```
Comment (mitza-oci) [inline dds/DCPS/WriteDataContainer.h:538]: needs allocator-enabled STL container

File: dds/DCPS/WriteDataContainer.cpp (hunk 2 above)
Comment (mitza-oci) [reply dds/DCPS/WriteDataContainer.cpp:1520]: discussed

File: dds/DCPS/WriteDataContainer.cpp [hunk 4]
```diff
@@ -1500,6 +1488,159 @@ WriteDataContainer::log_send_state_lists (OPENDDS_STRING description)
...
+      //const TimeDuration adjustment = deadline_period - deadline_period_;
+      const TimeDuration adjustment = deadline_period;
+      DeadlineMapType new_map;
+      for (PublicationInstanceMapType::iterator iter = instances_.begin();
+           iter != instances_.end();
+           ++iter) {
+        iter->second->deadline_ += adjustment;
```
Comment (simpsont-oci) [inline dds/DCPS/WriteDataContainer.cpp:1527]: So if the previous deadline_ was an hour in the future, but new deadline is 5 seconds, this looks like it's bumping the next deadline and next wait to 1hr + 5 seconds, which seems wrong?

File: dds/DCPS/WriteDataContainer.cpp [hunk 5]
```diff
@@ -1500,6 +1488,159 @@ WriteDataContainer::log_send_state_lists (OPENDDS_STRING description)
...
+        new_map.insert(std::make_pair(iter->second->deadline_, iter->second));
+      }
+      std::swap(new_map, deadline_map_);
+
+      if (!deadline_map_.empty()) {
+        deadline_task_->cancel();
+        deadline_task_->schedule(deadline_map_.begin()->first - MonotonicTimePoint::now());
```
Comment (simpsont-oci) [inline dds/DCPS/WriteDataContainer.cpp:1534]: I would maybe update the logic for calculating the next scheduled deadline similar between this line and how it's calculated above for a "new" deadline... i.e. have a temporary variable named deadline, just to make it clear what's going on here. Also be aware that sometimes canceling timers can take a while, so calculating "now" as close to function entry as possible is probably the desired behavior.

File: dds/DCPS/WriteDataContainer.h [hunk 6]
```diff
@@ -528,6 +531,28 @@ class OpenDDS_Dcps_Export WriteDataContainer : public PoolAllocationBase {
...
+  DeadlineMapType deadline_map_;
+
+  /// Lock for synchronization of @c status_ member.
+  ACE_Recursive_Thread_Mutex& deadline_status_lock_;
```
Comment (simpsont-oci) [inline dds/DCPS/WriteDataContainer.h:542]: Yuck. Understandable, but yuck.

File: dds/DCPS/WriteDataContainer.h [hunk 7]
```diff
@@ -528,6 +531,28 @@ class OpenDDS_Dcps_Export WriteDataContainer : public PoolAllocationBase {
...
+  /// Reverse lock used for releasing the @c status_lock_ listener upcall.
+  ACE_Reverse_Lock<ACE_Recursive_Thread_Mutex> deadline_reverse_status_lock_;
```
Comment (simpsont-oci) [inline dds/DCPS/WriteDataContainer.h:544]: Do we need this as a member? Can we just create these inline as we need them?

File: dds/DCPS/DataSampleElement.h [hunk 8]
```diff
@@ -31,6 +32,7 @@ typedef Cached_Allocator_With_Overflow<DataSampleElement, ACE_Null_Mutex>
 class TransportSendListener;
 struct PublicationInstance;
 typedef RcHandle<PublicationInstance> PublicationInstance_rch;
+typedef WeakRcHandle<PublicationInstance> PublicationInstance_wrch;
```
[the remark this answers is not in the document]
Comment (jrw972) [reply dds/DCPS/DataSampleElement.h:35]: I'll take it out see what happens.

File: dds/DCPS/WriteDataContainer.cpp (hunk 4 above)
Comment (jrw972) [reply dds/DCPS/WriteDataContainer.cpp:1527]: I agree.  I will change it so it resets all instances to now() + deadline_period.

File: dds/DCPS/WriteDataContainer.cpp (hunk 5 above)
Comment (jrw972) [reply dds/DCPS/WriteDataContainer.cpp:1534]: In this case, it is better to defer now() to as late as possible.  If I use now() upon function entry, then the relative deadline to schedule will be larger and the timer will fire after the deadline.

File: dds/DCPS/WriteDataContainer.h (hunk 7 above)
Comment (jrw972) [reply dds/DCPS/WriteDataContainer.h:544]: I'll inline it.

File: dds/DCPS/WriteDataContainer.h (hunk 6 above)
Comment (jrw972) [reply dds/DCPS/WriteDataContainer.h:542]: Yeah.  Passing around references to locks...what could possibly go wrong?

**Pull request merged**